Sources/Basic/TemporaryFile.swift

```diff
@@ -0,0 +1,172 @@
+    private let fd: Int32
+
+    /// FileHandle of the temporary file, can be used to read/write data.
+    public let fileHandle: NSFileHandle
+
+    /// Lazily initialized OutputByteStream object. Passed to the closure of write(write:) method for conveniently writing to file.
+    lazy private var stream = OutputByteStream()
+
+    /// Creates an instance of Temporary file. The temporary file will live on disk until the instance
+    /// goes out of scope.
+    ///
+    /// - Parameters:
+    ///     - dir: If specified the temporary file will be created in this directory otherwise enviornment variables
+    ///            TMPDIR, TEMP and TMP will be checked for a value (in that order). If none of the env variables are
+    ///            set, dir will be set to `/tmp/`.
+    ///     - prefix: The prefix to the temporary file name.
+    ///     - suffix: The suffix to the temporary file name.
+    ///
+    /// - Throws: TempFileError
+    public init(dir: String? = nil, prefix: String = "", suffix: String = "") throws {
+        self.suffix = suffix
+        self.prefix = prefix
+        // Determine in which directory to create the temporary file.
+        self.dir = TemporaryFile.determineTempDirectory(dir)
+        // Construct path to the temporary file.
+        let path = self.dir + prefix + "TemporaryFile.XXXXXX" + suffix
+
+        // Convert path to a C style string terminating with null char to be an valid input
+        // to mkstemps method. The XXXXXX in this string will be replaced by a random string
+        // which will be the actual path to the temporary file.
+        var template = ByteString(encodingAsUTF8: path).contents.map { Int8($0) } + [Int8(0)]
+
+        fd = libc.mkstemps(&template, Int32(suffix.characters.count))
+        // If mkstemps failed then throw error.
+        if fd == -1 { throw TempFileError(errno: errno) }
+
+        self.path = String(cString: template)
+        fileHandle = NSFileHandle(fileDescriptor: fd)
+    }
+
+    /// Convinence method for writing to the temporary file using OutputByteStream.
+    /// Note: This will write using the fileHandle and hence move the handle's filepointer.
+    ///
+    /// - Parameters:
+    ///     - write: Closure which provides the OutputByteStream object to write to.
+    ///
+    /// - Throws: SystemError
+    public func write(write: (OutputByteStream) -> Void) throws {
+        write(stream)
+        return try fputs(stream.bytes.contents, fileHandle)
+    }
+
+    /// Remove the temporary file before deallocating.
+    deinit { unlink(path) }
+
+    /// Create a default TemporaryFile object for duration of the closure passed to it.
+    ///
+    /// - Parameters:
+    ///     - body: Closure providing a default instance of TemporaryFile.
+    ///
+    /// - Throws: TempFileError
+    public static func withFile<T>(body: @noescape (TemporaryFile) throws -> T) throws -> T {
+        return try body(try TemporaryFile())
+    }
```

Comment:
This doesn't feel like the right place for this API to me, I don't think it adds sufficient value to be worth being on the `Basic` class.